lib/service/awsoidc.go

```diff
@@ -0,0 +1,414 @@
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// ownershipTags are used to identify if the ecs resources are managed by the
+	// teleport integration.
+	ownershipTags := map[string]string{
+		types.ClusterLabel:     updater.TeleportClusterName,
+		types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+		types.IntegrationLabel: integration.GetName(),
+	}
+
+	// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+	// If the lease cannot be acquired, the update is already being handled by another instance.
+	semLock, err := updater.AuthClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
+		SemaphoreKind: types.SemaphoreKindConnection,
+		SemaphoreName: fmt.Sprintf("update_deploy_service_agents_%s_%s", awsRegion, integration.GetName()),
+		MaxLeases:     1,
+		Expires:       updater.Clock.Now().Add(updateDeployAgentsInterval),
+		Holder:        "update_deploy_service_agents",
+	})
+	if err != nil {
+		if strings.Contains(err.Error(), teleport.MaxLeases) {
+			updater.Log.WithError(err).Debug("Deploy service agent update is already being processed.")
+			return nil
+		}
+		return trace.Wrap(err)
+	}
+	defer func() {
+		if err := updater.AuthClient.CancelSemaphoreLease(ctx, *semLock); err != nil {
+			updater.Log.WithError(err).Error("Failed to cancel semaphore lease.")
+		}
+	}()
+
+	updater.Log.Debugf("Updating Deploy Service Agents for integration %s in AWS region: %s", integration.GetName(), awsRegion)
+	if err := awsoidc.UpdateDeployServiceAgent(ctx, deployServiceClient, awsoidc.UpdateServiceRequest{
+		TeleportClusterName: updater.TeleportClusterName,
+		TeleportVersionTag:  teleportVersion,
+		OwnershipTags:       ownershipTags,
+	}); err != nil {
+
+		switch {
+		case trace.IsNotFound(err):
+			// The updater checks each integration/region combination, so
+			// there will be regions where there is no ECS cluster deployed
+			// for the integration.
+			updater.Log.WithError(err).Debugf("Integration %s does not manage any services within region %s.", integration.GetName(), awsRegion)
+			return nil
+		case trace.IsAccessDenied(awslib.ConvertIAMv2Error(trace.Unwrap(err))):
+			// The aws oidc role may lack permissions due to changes in teleport.
+			// In this situation users should be notified that they will need to
+			// re-run the deploy service iam configuration script and update the
+			// permissions.
+			updater.Log.WithError(err).Warning("Re-run deploy service configuration script to update permissions.")
+			if err := updater.ensureClusterAlert(ctx, integration, awsRegion); err != nil {
+				updater.Log.WithError(err).Warning("Failed to ensure cluster alert.")
+			}
+		}
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+// ensureClusterAlert ensures a cluster alert is created if deploy service permissions
+// need to be reconfigured.
+func (updater *DeployServiceUpdater) ensureClusterAlert(ctx context.Context, integration types.Integration, awsRegion string) error {
```

Comment:
I'm not sure a cluster alert is an appropriate tool for this kind of notification. I think we need to come up with a consistent way of how to attract appropriate users' attention to actions they need to perform, but that's out of scope of this issue.

I would keep cluster alerts out of the initial implementation, and I'll bring it up with the product team to see what we want to do going forward.